File: 2015/day18/day18.py
input_file = "/home/sunil/Desktop/repos/AoC/2015/day18/input.txt"
lights = [[0] * 100 for i in range(100)]

def processData(input_file, lights):
    with open(input_file, 'r') as file:
        for i, line in enumerate(file):
            for j, char in enumerate(line.strip()):
                if char == '#':
                    lights[i][j] = 1
                elif char == '.':
                    lights[i][j] = 0
    return lights



def animatePartOne(lights):
    next_state = [[0] * 100 for _ in range(100)] 
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]

    for i in range(100):
        for j in range(100):
            lightsOn = 0

            for dx, dy in directions:
                ni, nj = i + dx, j + dy
                if 0 <= ni < 100 and 0 <= nj < 100 and lights[ni][nj] == 1:
                    lightsOn += 1
            if (lights[i][j] == 1 and (lightsOn == 2 or lightsOn == 3)) or (lights[i][j] == 0 and lightsOn == 3):
                next_state[i][j] = 1
            else:
                next_state[i][j] = 0
    return next_state

def animatePartTwo(lights):
    next_state = [[0] * 100 for _ in range(100)] 
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1)]
    lights[0][0] = 1
    lights[0][99] = 1
    lights[99][0] = 1
    lights[99][99] = 1
    for i in range(100):
        for j in range(100):
            lightsOn = 0
            for dx, dy in directions:
                ni, nj = i + dx, j + dy
                if 0 <= ni < 100 and 0 <= nj < 100 and lights[ni][nj] == 1:
                    lightsOn += 1
            
            if (lights[i][j] == 1 and (lightsOn == 2 or lightsOn == 3)) or (lights[i][j] == 0 and lightsOn == 3):
                next_state[i][j] = 1
            else:
                next_state[i][j] = 0
    next_state[0][0] = 1
    next_state[0][99] = 1
    next_state[99][0] = 1
    next_state[99][99] = 1
    return next_state

def countLights(lights):
    count = 0
    for i in range(100):
        for j in range(100):
            if lights[i][j] == 1:
                count += 1
    return count

processedData = processData(input_file, lights)
animationPartOne = animatePartOne(processedData)
for i in range(99):
    animationPartOne = animatePartOne(animationPartOne)

countPartOne = countLights(animationPartOne)
print("Part One: " + str(countPartOne))

animationPartTwo = animatePartTwo(processedData)
for i in range(99):
    animationPartTwo = animatePartTwo(animationPartTwo)

countPartTwo = countLights(animationPartTwo)
print("Part Two: " + str(countPartTwo))

File: 2015/day18/day18.cpp
#include<iostream>
#include<fstream>
#include<string>
#include<sstream>
#define gridSize 100
using namespace std;

int processedData[gridSize][gridSize];
int tempData[gridSize][gridSize];
int nextGrid[gridSize][gridSize];

void processData(string line, int i){
    for (int j = 0; j < line.length(); j++){
        if (line[j] == '#'){
            processedData[i][j] = 1;
        } else if (line[j] == '.'){
            processedData[i][j] = 0;
        } 
    }
}

void animatePartOne(int array[gridSize][gridSize]) {
    int directions[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};

    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridSize; j++) {
            int lightsOn = 0;

            for (int k = 0; k < 8; k++) {
                int ni = i + directions[k][0];
                int nj = j + directions[k][1];
                if (ni >= 0 && ni < gridSize && nj >= 0 && nj < gridSize && array[ni][nj] == 1) {
                    lightsOn++;
                }
            }

            if ((array[i][j] == 1 && (lightsOn == 2 || lightsOn == 3)) || (array[i][j] == 0 && lightsOn == 3)) {
                nextGrid[i][j] = 1;
            } else {
                nextGrid[i][j] = 0;
            }
        }
    }
    for (int i = 0;i < 100; i++){
        for (int j = 0; j < 100; j++){
            tempData[i][j] = nextGrid[i][j];
        }
    }
}

void animatePartTwo(int array[gridSize][gridSize]) {
    int directions[8][2] = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}};

    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridSize; j++) {
            int lightsOn = 0;

            for (int k = 0; k < 8; k++) {
                int ni = i + directions[k][0];
                int nj = j + directions[k][1];
                if (ni >= 0 && ni < gridSize && nj >= 0 && nj < gridSize && array[ni][nj] == 1) {
                    lightsOn++;
                }
            }

            if ((array[i][j] == 1 && (lightsOn == 2 || lightsOn == 3)) || (array[i][j] == 0 && lightsOn == 3)) {
                nextGrid[i][j] = 1;
            } else {
                nextGrid[i][j] = 0;
            }
        }
    }

    for (int i = 0; i < gridSize; i++) {
        for (int j = 0; j < gridSize; j++) {
            tempData[i][j] = nextGrid[i][j];
        }
    }
    tempData[0][0] = 1;
    tempData[0][99] = 1;
    tempData[99][0] = 1;
    tempData[99][99] = 1;
}



int count(int array[gridSize][gridSize]){
    int total = 0;
    for (int i = 0;i < 100; i++){
        for (int j = 0; j < 100; j++){
            if (array[i][j] == 1){
                total++;
            }
        }
    }
    return total;
}

void executePartOne(){
    ifstream my_file("input.txt");
    if(!my_file) {
        cout << "Error: Unable to open the file." << endl;
        return;
    }
    string line;
    int i = 0;
    while (!my_file.eof()) {
        getline(my_file, line);
        processData(line, i);
        i++;
    }
    my_file.close();

    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            tempData[i][j] = processedData[i][j];
        }
    }
    for (int i = 0; i < 100; i++) {
        animatePartOne(tempData);
    }
    int total = count(nextGrid);
    cout << "Part One: " << total << endl;
}

void executePartTwo(){
    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            tempData[i][j] = processedData[i][j];
        }
    }
    for (int i = 0; i < 100; i++) {
        animatePartTwo(tempData);
    }
    int total = count(nextGrid);
    cout << "Part Two: " << total << endl;
}
int main(){
    executePartOne();
    for (int i = 0; i < 100; i++) {
        for (int j = 0; j < 100; j++) {
            tempData[i][j] = 0;
            nextGrid[i][j] = 0;
        }
    }
    executePartTwo();
    return 0;
}
